esphome/components/graph/__init__.py

```diff
@@ -0,0 +1,107 @@
+from esphome.components import display, sensor
+import esphome.config_validation as cv
+import esphome.codegen as cg
+from esphome.cpp_types import App
+from esphome.const import (
+    CONF_ID,
+    CONF_WIDTH,
+    CONF_SENSOR,
+    CONF_HEIGHT,
+    CONF_MIN_VALUE,
+    CONF_MAX_VALUE,
+    CONF_MIN_RANGE,
+    CONF_MAX_RANGE,
+    CONF_LINE_THICKNESS,
+    CONF_LINE_TYPE,
+    CONF_UPDATE_INTERVAL,
+    CONF_X_GRID,
+    CONF_Y_GRID,
+    CONF_BORDER,
+    CONF_AXES,
+)
+
+CODEOWNERS = ["@synco"]
+
+DEPENDENCIES = ["display"]
+MULTI_CONF = True
+
+LineType = display.display_ns.enum("LineType")
+LINE_TYPE = {
+    "SOLID": LineType.LINE_TYPE_SOLID,
+    "DOTTED": LineType.LINE_TYPE_DOTTED,
+    "DASHED": LineType.LINE_TYPE_DASHED,
+}
+
+Graph_ = display.display_ns.class_("Graph")
+graph_ns = cg.esphome_ns.namespace("graph")
+GraphAxes = graph_ns.class_("GraphAxes")
+GraphAxesPtr = GraphAxes.operator("ptr")
+
+GRAPH_BASIC_SCHEMA = cv.Schema(
+    {
+        cv.Required(CONF_ID): cv.declare_id(Graph_),
+        cv.Required(CONF_WIDTH): cv.positive_int,
+        cv.Required(CONF_HEIGHT): cv.positive_int,
+    }
+).extend(cv.polling_component_schema("10s"))
+
+CONFIG_SCHEMA = GRAPH_BASIC_SCHEMA.extend(
+    {
+        cv.Optional(CONF_SENSOR): cv.use_id(sensor.Sensor),
+        cv.Optional(CONF_MIN_VALUE, default=float("nan")): cv.float_,
+        cv.Optional(CONF_MAX_VALUE, default=float("nan")): cv.float_,
+        cv.Optional(CONF_MIN_RANGE, default=float("nan")): cv.float_,
+        cv.Optional(CONF_MAX_RANGE, default=float("nan")): cv.float_,
+        cv.Optional(CONF_LINE_THICKNESS, default=3): cv.positive_int,
+        cv.Optional(CONF_LINE_TYPE, default="SOLID"): cv.enum(LINE_TYPE, upper=True),
+        cv.Optional(CONF_X_GRID, default=float("nan")): cv.float_,
+        cv.Optional(CONF_Y_GRID, default=float("nan")): cv.float_,
```

Comment:
(1/3) Don't use `nan` defaults, just use no default and check if the key is in the config, and initialize the variables to NAN in c++ (see other comments/suggestions)
```suggestion
        cv.Optional(CONF_MIN_VALUE): cv.float_,
        cv.Optional(CONF_MAX_VALUE): cv.float_,
        cv.Optional(CONF_MIN_RANGE): cv.float_,
        cv.Optional(CONF_MAX_RANGE): cv.float_,
        cv.Optional(CONF_LINE_THICKNESS, default=3): cv.positive_int,
        cv.Optional(CONF_LINE_TYPE, default="SOLID"): cv.enum(LINE_TYPE, upper=True),
        cv.Optional(CONF_X_GRID): cv.float_,
        cv.Optional(CONF_Y_GRID): cv.float_,
```

Reply:
Is that:
a) no defaults in *.py, (happy - with that)
b) no float("nan") yet other defaults ok. (I find that confusing)
c) other

Reply:
I like the idea of using a sensor callback, yet suspect using a callback on sensor state would cause issues -as sensor poll rate is most likely different to the graph sampling rate.

Reply:
> I like the idea of using a sensor callback, yet suspect using a callback on sensor state would cause issues -as sensor poll rate is most likely different to the graph sampling rate.

I agree we should use sensor callback, then we also don't need the `update_interval` option because it just uses what the sensor has itself.

Yes the sensor/display refresh rates won't match, so I think we should add an option to control the x-scaling (essentially, how long in time units the history should be).

Reply:
Done